### DISCUSSION TITLE
PolarFire SoC USB Linux driver, getting error "cannot open device /dev/video0 exiting"

### DISCUSSION THREAD

**Discussion by imaginations635** [polarfire-soc/polarfire-soc-documentation#267] [Linux drivers for PolarFire Soc] [answered]
hello,
while connecting webcam to ICICLE kit getting error  "cannot open device /dev/video0 exiting"
command  $ v4l2-ctl --device /dev/video0 --set-fmt-video=width=640,height=480,pixelformat=MJPG --stream-mmap=3 --stream-count=100 --stream-to=stream.vid

**Comment by hughbreslin** [comment]
This appears to be a usage question, this should be a discussion item as per the issue template and contributing guidelines.

**Comment by hughbreslin** [comment, marked as answer]
Hi @imaginations635 can you please run `lsusb` and see if your camera is being detected?

If it is try running `lsusb -d [ID] -v |grep -e Class -e Config` and see the `bNumConfigurations` value? for example `lsusb -d 046d:085c -v |grep -e Class -e Config`.

**Reply by imaginations635** [reply]
Hi @hughbreslin 
results are -
root@icicle-kit-es:~# lsusb
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
root@icicle-kit-es:~# lsusb -d 1d6b:0002 -v |grep -e Class -e config
  bDeviceClass            9 Hub
  bDeviceSubClass         0
  bInterfaceClass         9 Hub
  bInterfaceSubClass      0

**Reply by hughbreslin** [reply]
Just to confirm the camera is connected over usb? It doesn’t appear to have been detected on the usb hub

**Reply by imaginations635** [reply]
yes it is directly connected on J16 port via USB OTG cable

**Reply by hughbreslin** [reply]
Hmm odd, can you confirm the setting of J15 and J17 are correct for otg? You can find them [here](https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/boards/mpfs-icicle-kit-es/icicle-kit-user-guide/icicle-kit-user-guide.md#jumpers)

**Reply by imaginations635** [reply]
yes correct both J15 & J17 are closed , I'm using Logitech C270 webcam. 
I also included this packages in yocto image.
yavta \
libuvc \
gstd \
gstreamer1.0-plugins-good \
v4l-utils \

**Reply by hughbreslin** [reply]
Hmm ok I will need to think about this - I’ll test when I’m at my desk tomorrow with a webcam and see if it’s coming up on my end. I don’t think this is getting far enough to be a v4l issue - I don’t think the camera is being detected at all. One thing you could test is connecting a different peripheral (e.g a thumb drive) and see if that’s detected on the hub.

**Reply by hughbreslin** [reply]
Two things to check actually, are you using the minimal image or the dev cli image? Could you try with the dev cli image as this should have all the packages needed by default. The thumbdrive will be an interesting test - this could be a power issue where the kit can't deliver the power for the camera, if you have a powered USB hub you could try with that and see if it makes a difference too. I'll still give this a try tomorrow. :)

**Reply by imaginations635** [reply]
Hi @hughbreslin 
I connected a pendrive to the board and it Successfully detected without USB hub. But the camera was not detected by board. 
What about the test result you tested webcam on board.

**Reply by hughbreslin** [reply]
Hi @imaginations635 

I just completed it - my kit seems to be detecting a camera ok:

```bash
root@icicle-kit-es:~# lsusb
Bus 001 Device 002: ID 05a3:9331 ARC International Camera
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
root@icicle-kit-es:~#
```

When I connect the camera to the kit while its running I get a vbus error but having it connected from boot it comes up and is detected ok - I'm only using our minimal build so I don't have all of the packages required to test it further but it showing up from `lsusb` is encouraging enough that the USB is working.

When I look at the `dmesg` output I can see the following:

```bash
[    9.894910] usb 1-1: new high-speed USB device number 2 using musb-hdrc
[   10.123954] mc: Linux media interface: v0.10
[   10.310930] videodev: Linux video capture interface: v2.00
[   10.404968] usb 1-1: Found UVC 1.00 device HD Web Camera (05a3:9331)
[   10.448463] usb 1-1: Failed to query (GET_INFO) UVC control 9 on unit 2: -32 (exp. 1).
[   10.479669] usb 1-1: Failed to query (GET_INFO) UVC control 4 on unit 2: -32 (exp. 1).
[   10.505618] usb 1-1: Failed to query (GET_INFO) UVC control 11 on unit 2: -32 (exp. 1).
[   10.538314] input: HD Web Camera: HD Web Camera as /devices/platform/soc/20201000.usb/musb-hdrc.1.auto/usb1/1-1/1-1:1.0/input/input0
[   10.550576] usbcore: registered new interface driver uvcvideo
``` 

can you see anything similar? Or if you run `dmesg` is there a "VBUS_ERROR" warning anywhere? This could indicate that power is indeed the problem and an externally powered hub will help. I have an externally powered hub here but its pretty much EOL and causes BSOD on my Windows machine and isn't being detected by Linux so I don't trust it for a reliable result.

**Reply by imaginations635** [reply]
no I'm getting this errors, please share download link of build you are using or is it possible I can connect to server directly without flashing new build  
[    7.870897] usb 1-1: device not accepting address 4, error -71
[    8.426933] usb 1-1: device not accepting address 5, error -71
[    8.435663] usb usb1-port1: unable to enumerate USB device

**Reply by hughbreslin** [reply]
I'm using the core-image-minimal-dev-icicle-kit-es-20221124145441.rootfs.wic.gz image available from this [link](https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp/releases)

**Reply by hughbreslin** [reply]
Hey @imaginations635 any update on this? :) Just checking if we can mark it as answered
